### PR TITLE
WIP: overwrite update, insert, delete

### PR DESCRIPTION
**Pull request by areski**
In relation to https://github.com/aesmail/kaffy/issues/85
I found that the suggested solution lack of flexibility because it will mean that one will need to refactor a lot of code in his phoenix project rather than using their own existing functions.

This can become very complex, in my example, we upload & convert images, insert in the DB and then also generate thumbnail and insert them too, for reference: https://github.com/areski/ex-chitchat/blob/master/lib/chit_chat_web/controllers/upload_controller.ex#L33

I think there is also many situation where this might happen.

Therefore, in the same way we have many callback (before/after/create etc...) we could introduce a new method that can be defined in the admin module: `overwritten_create`, along with some other `overwritten_update` & `overwritten_delete`

`overwritten_create` will allow you to bypass Kaffy repo insert and write your own.

Let me know what you think about it, I will continue the WIP if you think that's acceptable.

### CODE REVIEW

**Comment by aesmail** [comment]
I think this goes well with Kaffy's philosophy of having sensible and customizable defaults.

I would however change the function names from `overwritten_create` to simply `create`. And now thinking about it, maybe we should use ecto's terms to be clearer:

`before_create` -> `before_insert`
`after_create` -> `after_insert`
and have an `insert` function that, if present, will be used instead of Kaffy's own function.

What do you think?

**Comment by areski** [comment]
:100: for this!

I guess it will simply our life if we don't have to reinvent our own terminology, I personally spend too much time trying to find names for things without ever being satisfy with the decision :smile: 

On a side note, I see that  the callbacks`before_insert` etc.. have been depreciated
https://hexdocs.pm/ecto/1.1.0/Ecto.Model.Callbacks.html
It doesn't really matter we could still use those names.

**Comment by aesmail** [comment]
> I personally spend too much time trying to find names

Yup, that's one of my problems too.

> callbacks `before_insert` etc.. have been depreciated

I actually miss those functions. I use them a lot with my rails applications. When ecto v2 shipped, there were no more callbacks. I understand the ecto team's decision to remove them and I realize the pitfalls of having callbacks, but I believe it's up to the developer to use them wisely and it's nice to have this option available when needed. Maybe we can emphasize that using them is discouraged or they shouldn't be the first solution to look for? I'm not sure.

**Comment by areski** [comment]
I have never used Rails... what would be the pitfalls of using those, increase complexity in written code?

**Comment by aesmail** [comment]
Yes, it could become hard to follow the logic of the code since you can't really see the callback functions (or methods) being called anywhere in your code. They're called implicitly.

I have been bitten a few times by this from my own code where some models weren't being saved and I didn't know why (turns out callbacks were to blame). But they're so convenient when you need them that I'm willing to have the option at least.

**Comment by aesmail** [comment]
cc @zenbaku 
just an FYI, keeping you in the loop of what's coming so there's no overlap.

**Comment by areski** [comment]
It's working nicely for me now, let me know what you think?
I will rename the other method in a different commit

**Comment by areski** [comment]
Documentation added!

File: lib/kaffy/resource_callbacks.ex [hunk 1]
```diff
@@ -60,20 +84,42 @@ defmodule Kaffy.ResourceCallbacks do
...
       false
     )
   end
 
-  defp after_create(conn, resource, entry) do
+  defp insert(conn, resource, changeset) do
+    Utils.get_assigned_value_or_default(
+      resource,
+      :insert,
+      {:not_found},
```
Comment (aesmail) [inline lib/kaffy/resource_callbacks.ex:113]: Any reason this is a tuple and not simply `:not_found`?

**Comment by areski** [comment]
I made a correction as recommended!
Thanks for the review :smiley:
